ir/src/constraints/graph.rs

```diff
@@ -108,13 +108,18 @@ impl AlgebraicGraph {
                 default_domain,
             ),
             Expression::Rand(index) => {
-                // The constraint target for random values defaults to the second (auxiliary) trace
-                // segment.
+                // constraint target for random values defaults to the second trace segment.
                 // TODO: make this more general, so random values from further trace segments can be
                 // used. This requires having a way to describe different sets of randomness in
                 // the AirScript syntax.
-                let node_index = self.insert_op(Operation::RandomValue(*index));
-                Ok(ExprDetails::new(node_index, AUX_SEGMENT, default_domain))
+                if *index >= symbol_table.num_random_values() as usize {
+                    return Err(SemanticError::IndexOutOfRange(format!(
+                        "Random value index {} is greater than or equal to the total number of random values ({}).", 
+                        index,
+                        symbol_table.num_random_values()
+                    )));
+                }
+                self.insert_random_value(*index, default_domain, AUX_SEGMENT)
```

Comment:
nit: I'd prefer to pass the symbol table to the `insert_random_values` method as well so that the validity check and error can be moved inside that method  (`if *index >= symbol_table.num_random_values()`). That will keep this `insert_expr` method cleaner & match the pattern of the other branches